Windows manifest improvements

FranciscoPombal opened the pull request:
Based on the discussion in https://github.com/qbittorrent/qBittorrent/issues/13748. I think there are also other relevant issues that _may_ get fixed by this (such as those related to multi-monitor GUI troubles).

Won't mark this as draft since it is ready code-wise, but let's first give team members and users the chance to try out the CI builds to confirm that at least see no regressions before merging.

FranciscoPombal commented:
ping @xavier2k6

xavier2k6 commented:
May be insignificant but why not use:
`<ws2:longPathAware xmlns:ws2="http://schemas.microsoft.com/SMI/2016/WindowsSettings">true</ws2:longPathAware>`?

As per:
https://docs.microsoft.com/en-au/windows/win32/sbscs/application-manifests#longpathaware

Just to be more consistent with what is recommended by Microsoft.


I have also used `Segement Heap` in manifest file too previously.......may increase CPU Utilization....

FranciscoPombal commented:
@xavier2k6 

> May be insignificant but why not use:
> `<ws2:longPathAware xmlns:ws2="http://schemas.microsoft.com/SMI/2016/WindowsSettings">true</ws2:longPathAware>`?
> 
> As per:
> https://docs.microsoft.com/en-au/windows/win32/sbscs/application-manifests#longpathaware
> 
> Just to be more consistent with what is recommended by Microsoft.

No need, all attributes explicitly declare the namespace anyway, no need for one with the `ws2` prefix. If anything, I think it's Microsoft's own documentation that is inconsistent in this regard.

> I have also used `Segement Heap` in manifest file too previously.......may increase CPU Utilization....

Can you better quantify/substantiate this? I haven't observed it.

> EDIT:
> 
> Also to point out with the `PMV2` addition, we probably don't now need to pragmatically code for #13675

I assume you mean "programmatically"? Yeah, I'm also under the impression that this is the only required change. Other than this, we just have to hope that Qt itself also has proper support in place.

xavier2k6 commented:
>No need, I have explicitly declared the namespace for all attributes, no need for one with the ws2 prefix. If anything, I think it's Microsoft's own documentation that is inconsistent in this regard.

Fair enough.

>Can you better quantify/substantiate this? I haven't observed it.

I haven't observed it myself, but it's just that Google Chrome moved over to segment heap previously & they observed an increase in CPU.......it effected a benchmark I think (will have to look for link) but they removed it again.

merely put it there as a disclaimer........just in case 😄 

>I assume you mean "programmatically"?

Ya, that was a typo! 🙄

FranciscoPombal commented:
@xavier2k6 

> I haven't observed it myself, but it's just that Google Chrome moved over to segment heap previously & they observed an increase in CPU.......it effected a benchmark I think (will have to look for link) but they removed it again.
> 
> merely put it there as a disclaimer........just in case smile

Fair enough, but Google Chrome is only one piece of software, and an extremely complex one at that. It's quite different from qBittorrent, so I don't think such observations are too relevant. Microsoft recommends it as a general improvement, so unless a serious regression is observed, we should enable it for systems that can take advantage of it.

xavier2k6 commented:
>Fair enough, but Google Chrome is only one piece of software, and an extremely complex one at that. It's quite different from qBittorrent, so I don't think such observations are too relevant. Microsoft recommends it as a general improvement, so unless a serious regression is observed, we should enable it for systems that can take advantage of it.

Agree

FranciscoPombal commented:
> Commit titles are incorrect.

@glassez 

Is it OK now? If not, please post suggestions.

glassez commented:
> Is it OK now?

Ok. Except typo in the second commit.

FranciscoPombal commented:
@glassez 

> Ok. Except typo in the second commit.

Typo? Haven't you heard of `Windwos`, the brand new OS by Microsoft, successor to Windows 10 and meant to compete with Fuchsia?

Done.

glassez commented:
> Haven't you heard of `Windwos`, the brand new OS by Microsoft, successor to Windows 10 and meant to compete with Fuchsia?

😊

xavier2k6 commented:
>the brand new OS by Microsoft, successor to Windows 10**X**

I thought it was being re-branded to `Win dose` 🤣

glassez commented on src/qbittorrent.exe.manifest:
```diff
@@ -1,45 +1,39 @@
+    <!-- Set important attributes for better application behavior on modern systems -->
+    <asmv3:application>
+        <asmv3:windowsSettings>
+            <activeCodePage xmlns="http://schemas.microsoft.com/SMI/2019/WindowsSettings">UTF-8</activeCodePage>
```
Seems it still doesn't affect qBittorrent.
Our own code uses only Unicode variant of Win32 API.
I dumped official Qt binaries, and it confirmed my belief that it doesn't use ANSI Win32 API either. In addition, it is really unprofitable for Qt to use ANSI Win32 API (even with UTF8 code page), since it itself uses UTF16 to encode strings, so it would require frequent transcoding between UTF8 (or any other 1 byte encoding) and UTF16.

FranciscoPombal replied:
@glassez still, since there is no downside, might as well leave it like this since it is the recommended practice, no?

Chocobo1 replied:
>since it is the recommended practice, no?

I think not, IMO the real recommended practice would be MS set it as the default, not telling/urging users to opt-in.

qbt has always worked fine without it and adding this directive which we have no need and no effect sounds like redundant (or technical debt) for me.
Same viewpoint on other directives you added, you should propose them **only if** they have **real** impacts/effects, not because they are new or shiny/trendy or do some *suspected* fix.
At least separate them to its own commits and add appropriate comments/rational for them.

FranciscoPombal replied:
@Chocobo1

OK, I will separate each into separate commits.

>I think not, IMO the real recommended practice would be MS set it as the default, not telling/suggesting users to opt-in.

This logic is flawed. Not sure what the defaults for manifest generation in VS actually are, but just because some "old" value is the default does not mean it is the recommended practice, or that the "new" values are reserved for "bleeding edge" cases, especially in ecosystems where backwards compatibility is a core design goal, such as is the case for Windows. The docs make it pretty clear what the recommended practices are. If you don't want these changes, enjoy receiving endless bug reports about broken multi-monitor support due to not setting PerMonitorV2, which is a prerequisite for fixing it on modern systems.

About each individual addition:

- `dpiAware`/`dpiAwareness`: This has been discussed at length in the past, especially with @xavier2k6 (https://github.com/qbittorrent/qBittorrent/projects/2#column-12115743). These changes are required for proper multi-monitor support.
- `activeCodePage`: This is recommended by the docs, and only "affects" Windows 10 >= 1903 anyway (https://docs.microsoft.com/en-au/windows/win32/sbscs/application-manifests#activecodepage, https://docs.microsoft.com/en-us/windows/uwp/design/globalizing/use-utf8-code-page). Not keeping up with these recommendations is what accrues technical debt, not adopting them. Adopting such changes prevents potential issues down the line, if it does not explicitly "fix" anything now.
- `heapType`: similar to the previous point. This is the new modern way (https://docs.microsoft.com/en-au/windows/win32/sbscs/application-manifests#heaptype), and only applies to Windows 10 >= 2004.

> Also "Reorganize and reformat Windows manifest" commit looks meaningless to me, I vote to drop it.

It simply formats the file with 4 space indentation, similar to the main source files, which is more consistent, and moves the `<compatibility>` stanza to the top, which is the most important one. I don't think these are bad changes. Are you sure?

glassez replied:
>`activeCodePage`: This is recommended by the docs

Seems "you heard bells, but you didn't know where the sound was coming from".
Where exactly in the documentation is it strictly recommended for unconditional use in all cases? It just mentions that this might be a good idea in general, in case you're dealing with UTF-8 encoded strings, and the application uses ANSI API.
But this does not apply well to qBittorrent (as, indeed, to any Qt application). Although we're dealing with UTF-8 encoded strings passed from some outer sources we still have convert them to QString (i.e. in UTF-16) in the vast majority of cases. So why we need to convert them back to UTF-8 when calling Win32 API instead of using its Unicode (UTF-16) variant? Especially considering the fact that Windows operates with UTF-16 natively.
If `activeCodePage` doesn't affect us, why should we clog our Manifest with it?
>Adopting such changes prevents potential issues down the line, if it does not explicitly "fix" anything now.

`activeCodePage` option can neither prevent nor fix anything by itself. It just allows you to use UTF-8 (or other code page) in your application regardless of system one (if you use ANSI Win32 API).

Chocobo1 replied:
>This logic is flawed. 

Sorry, but it works well until proven otherwise.

>but just because some "old" value is the default does not mean it is the recommended practice

I never said that.

>or that the "new" values are reserved for "bleeding edge" cases

Yes it is. If it is not then why don't they enable it by default and benefit everyone?

>If you don't want these changes, enjoy...

Whatever you say.... try show some real confirmations from users, you have proven nothing yet.

>I don't think these are bad changes. Are you sure?

Yes, I won't be pointing out if I don't care.

FranciscoPombal replied:
@Chocobo1 

> Sorry, but it works well until proven otherwise.

The logic of the statement "I think not, IMO the real recommended practice would be MS set it as the default, not telling/suggesting users to opt-in." is flawed at fundamental theoretical level and goes against what is written in the docs. There is nothing to be "proven otherwise". They _do_ recommend switching, they just don't do it _by default_ because of backwards compatibility/legacy users. Similar case to `/Zc:__cplusplus`.

> I never said that.

It is the logical consequence of this, which you did say:

> I think not, IMO the real recommended practice would be MS set it as the default, not telling/suggesting users to opt-in.

---

> Yes it is. If it is not then why don't they enable it by default and benefit everyone?

Because of backwards compatibility/legacy users in the wider Windows ecosystem. Read my previous post again. We've been over this in previous instances, such as with `/Zc:__cplusplus`.

> Whatever you say.... try show some real confirmations from users that your proposed changes worked, until now you have proven nothing yet.

From reading the documentation it should be obvious to you that these changes are a net benefit, _regardless of whether you can already see a "confirmation" right now_. But, since you insist so much: https://github.com/qbittorrent/qBittorrent/issues/15019#issuecomment-851013929.

Now the burden of proof is on you to prove how any of this is a regression or breaks something.

> And don't forget to put them in the commit comment.

:+1:

FranciscoPombal replied:
Fixed.

xavier2k6 commented:
I encountered an exception previously that was rectified by libtorrent through UTF-8 sanitization, so using UTF-8 overall may not be actually a bad thing?!

ref.:
#11661
https://github.com/arvidn/libtorrent/issues/4345
https://github.com/arvidn/libtorrent/pull/4922

xavier2k6 commented:
"Reorganize and reformat Windows manifest" commit looks aesthetically better to me & it's easily more readable.

In relation to `dpiaware/dpiawareness` addition to manifest file - has fixed an issue & will potentially fix others
ref.: #15019

`dpiaware/dpiawareness` change is part 1 of about roughly a 3 part solution for current GUI issues in relation to HiDPI/Scaling

`UTF-8 Code Page` is a subjective change & although it may not benefit `locale` issues - it may benefit in other issues (Yet, to be confirmed)

`segment heap` could potentially aid in issues about memory use on Windows 10 >=2004/(20H1)

We also have to take in to consideration that - the changeover to Qt6 in the future requires a drop in older OS's/certain Windows 10 Build/Versions.

glassez commented:
> `UTF-8 Code Page` is a subjective change & although it may not benefit `locale` issues - it may benefit in other issues

It feels like you're not even trying to answer the basic question: how can a thing that isn't being used may benefit?

>(Yet, to be confirmed)

Well, if you can do that, I'll be happy to admit I'm wrong. But until then, please use it locally so as not to mislead anyone.

xavier2k6 commented:
>how can a thing that isn't being used may benefit?

All strings in .torrent files are UTF-8 encoded, or at least they are treated as UTF-8. No?
libtorrent verifies all encoding by decoding all characters and re-encodes them, it also validates UTF-8 encoding of client version strings from peers to prevent attacks/crashes.

>please use it locally so as not to mislead anyone.

I don't see where I misled anyone?? (It certainly isn't my attention to mislead)

In another thread - I simply showed my `local` manifest file, I didn't claim that - hey this is the ultimate elixir......roll up roll up, this will fix all your ailments.

I did however suggest elsewhere that the `dpiaware/dpiawarness` would help & I've shown proof of where that has helped in this PR.

glassez commented:
> > how can a thing that isn't being used may benefit?
> 
> All strings in .torrent files are UTF-8 encoded, or at least they are treated as UTF-8. No?
> libtorrent verifies all encoding by decoding all characters and re-encodes them, it also validates UTF-8 encoding of client version strings from peers to prevent attacks/crashes.

Please re-read https://github.com/qbittorrent/qBittorrent/pull/15020#discussion_r640531466.

FranciscoPombal commented:
@glassez  @Chocobo1

The DPI stuff has (predictably) been confirmed to work: https://github.com/qbittorrent/qBittorrent/issues/15019#issuecomment-851013929.
See also https://en.wikipedia.org/wiki/Doubting_Thomas.

sledgehammer999 commented:
About the UTF-8 controversy:
Terms used:
- `WINAPI` = Windows API
- `WCHAR` = is usually `wchar_t`
- ANSI WINAPI = WINAPI names that end in `A`. These take a string of type `char*` and expect it to be in the current code page encoding.
- UNICODE WINAPI = WINAPI names that end in 'W'. These take a string of type `wchar_t*` and expect it to be in utf-16 encoding.

Up until microsoft introduced the utf-8 manifest entry (Windows 10 Version 1903) programmers that wanted to use the ANSI WINAPI detected the used ANSI code page with [GetACP()](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getacp) and then converted their strings from utf-16 (wide string) to the ansi codepage using MultiByteToWideChar()/WideCharToMultiByte(). These 2 functions also changed the type used to store the string between `char*` and `wchar_t*`.

I suspect that when `activeCodePage=UTF-8` is used GetACP()/MultiByteToWideChar()/WideCharToMultiByte() will transparently keep working correctly. `GetACP()` will return utf-8 and the other 2 functions with convert to/from utf-8 accordingly.

qBittorrent itself doesn't use ANSI WINAPI directly. The other dependencies may use it or not, but those are probably covered by the above observation. Aka they shouldn't fail. They should transparently work. But I would expect that Qt and libtorrent use the UNICODE WINAPI already.
<br><br>

It is another matter entirely if this is actually useful in our context and it if it will solve any problems. This option seems like a convenience to cross-platform programmers. In linux the system API calls are usually in utf-8 stored in `char*`. Programs written with this assumption will need to have string conversions from/to utf-16 (using `whar_t*` as type) in order to use the UNICODE WINAPI. But if they can instead declare `activeCodePage=UTF-8` they don't need to change their string type (it will remain `char*`) nor do string encoding conversions (they can use the ANSI WINAPI instead).
But as I said it doesn't affect us. We interface with the underlying OS through cross-platform APIs (mainly Qt, boost, libtorrent). If those libs have a bug it will be unlikely that it will be fixed because we changed the activecodepage.

IMO, this seems like a step backwards for MS/WINDOWS. The ANSI WINAPI was considered kinda deprecated. But with this new ability they give it a new breath of life.

**To non-programmers**: It doesn't matter if external files (eg .torrent files) are stored in utf-8. They will work regardless of the WINAPI used (unless of course you use it incorrectly already).

TLDR: This manifest option should probably be dropped.

FranciscoPombal commented:
@sledgehammer999 @Chocobo1 @sledgehammer999 

Fine, I dropped the code page change. Commits reworded with all the necessary documentation and fixed up.

I have no interest in arguing about its inclusion in this PR anymore, let's just get this merged and over with.

---

/offtopic

>qBittorrent itself doesn't use ANSI WINAPI directly. The other dependencies may use it or not, but those are probably covered by the above observation. Aka they shouldn't fail. They should transparently work. But I would expect that Qt and libtorrent use the UNICODE WINAPI already.

>But as I said it doesn't affect us.

But if do happen to use it in the future, it doesn't hurt to be protected from shooting ourselves in foot. It doesn't affect us _right now_, but this is just gratuitously leaving a potential issue open for later. I will never understand the logic of not adopting a change with no present or future downside, when there is a potential future downside to keeping things the way they are. This is simple defensive programming.

> It is another matter entirely if this is actually useful in our context and it if it will solve any problems. This option seems like a convenience to cross-platform programmers.

qBittorrent is a cross-platform application, which would make us "cross-platform programmers" in this context, would it not? We should not turn down conveniences that exist for our sake...

>It is another matter entirely if this is actually useful in our context and it if it will solve any problems. This option seems like a convenience to cross-platform programmers. In linux the system API calls are usually in utf-8 stored in `char*`. Programs written with this assumption will need to have string conversions from/to utf-16 (using `whar_t*` as type) in order to use the UNICODE WINAPI. But if they can instead declare `activeCodePage=UTF-8` they don't need to change their string type (it will remain `char*`) nor do string encoding conversions (they can use the ANSI WINAPI instead).

This is a _good thing_, no more worrying about types depending on the platform, so, how can you come to the following conclusion?

> IMO, this seems like a step backwards for MS/WINDOWS. The ANSI WINAPI was considered kinda deprecated. But with this new ability they give it a new breath of life.

Deprecated in favor of a sub-par crutch, which is thankfully being done away with in favor of the real solution: proper support for Unicode via UTF-8.

There's even this quote from the Microsoft docs, which emphasizes the fact that they're now (correctly) emphasizing the use of `-A` APIs together with `activeCodePage=UTF-8`, instead of any other (worse) solution (https://docs.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page): 

> Until recently, Windows has emphasized "Unicode" -W variants over -A APIs. However, recent releases have used the ANSI code page and -A APIs as a means to introduce UTF-8 support to apps. If the ANSI code page is configured for UTF-8, -A APIs operate in UTF-8.

sledgehammer999 commented:
>But if do happen to use it in the future, it doesn't hurt to be protected

1. IIRC we have an (implicit) rule that we won't use the ANSI WINAPI in our code. Only the UNICODE WINAPI
2. It doesn't "protect" us from anything. It is just a convenience option to make it easier at writing code without converting between data types.

glassez commented:
@FranciscoPombal 
https://github.com/qbittorrent/qBittorrent/pull/15020#issuecomment-869874697 demonstrates that you either did not read (or did not understand) my comments above, or you do not understand the key aspects of this issue at all. You are again clinging to these abstract recommendations, completely not applying them to real use (namely in qBittorrent).

FranciscoPombal commented:
@sledgehammer999 

> 1. IIRC we have an (implicit) rule that we won't use the ANSI WINAPI in our code. Only the UNICODE WINAPI

To be clear: I am not criticizing or questioning the "original" decision of only using the W- APIs. That decision was made at a certain time, and under certain circumstances. Times and circumstances change, and thus past decisions should be reevaluated accordingly. It does not mean that it was a "bad" decision in the past. But now it goes against first-party recommendations in the official docs.

> 2. It doesn't "protect" us from anything. It is just a convenience option to make it easier at writing code without converting between data types.

It protects us from the inconvenience of having to convert between data types when writing code if we ever really need to use (I can imagine certain possible scenarios, such as when writing platform-specific code for stacktraces, for instance, or something like that).

@glassez 

> You are again clinging to these abstract recommendations, completely not applying them to real use (namely in qBittorrent).

Unsure what you mean by "abstract recommendations", these are very concrete recommendations from the official docs (which even explain the reasoning and historical context for the present recommendation). I am of the opinion that, unless we have a _very good reason™_ to go against the official docs, we should follow them. "It [presently] doesn't affect us" is not a good reason IMO.

Chocobo1 commented:
In git commit message:
> heapType:
https://docs.microsoft.com/en-au/windows/win32/sbscs/application-manifests#heaptype
>
>Setting to "SegmentHeap" enables the new heap implementation recommended by Microsoft.

I didn't find in the provided link that MS **recommended** it.

glassez commented:
> I am not criticizing or questioning the "original" decision of only using the W- APIs. That decision was made at a certain time, and under certain circumstances. Times and circumstances change, and thus past decisions should be reevaluated accordingly.

Circumstances are still the same. Some of them are described there (if you still haven't noticed this comment before):
https://github.com/qbittorrent/qBittorrent/pull/15020#discussion_r640038846

No one belittles the advantages of utf-8. It is really good for storing and transferring strings. But it is clearly inconvenient for processing (e.g. just read utf-8 data into `std::string` and try to obtain its size in characters).
Fortunately, we do not need to puzzle over this, since we use `QString`. So all we need to do with utf-8 is load the input data into a `QString`, i.e. convert it to utf-16, do whatever we want to do with it, and convert it to utf-8 when we pass it to where utf-8 is required. But **you don't need** to convert them to utf-8 to call Win32 API functions (even if Win32 API supports utf-8, and there are some recommendations to use it, that you don't understand the context of), because **Windows uses utf-16 natively**.

FranciscoPombal commented:
@Chocobo1 

> In git commit message:

> I didn't find in the provided link that MS **recommended** it. I would suggest you revise the wording.

This is a little obtuse. You could ask Microsoft to be more explicit in their docs, and write "this is the new recommended heap type where available". But just because it isn't written that way word-for-word right now, doesn't mean my assertion is wrong. It's more than obvious that this is the new recommendation (to be fair, I also wished Microsoft stated things like these more explicitly in the docs, documentation should not be left open to interpretation and "reading between the lines"):

- Quote from the docs: 
  > Segment heap is a modern heap implementation that will generally reduce your overall memory usage.

- Microsoft is moving (and has already moved some of) their own programs (including critical WIndows bits) to the new implementation:

    - https://blogs.windows.com/msedgedev/2020/06/17/improving-memory-usage/. It even references the docs page (emphasis mine):
        > Want to learn more about how developers can take advantage of the expanded availability of segment heap memory **improvements**? [Please visit this page for the latest developer information](https://docs.microsoft.com/en-us/windows/win32/sbscs/application-manifests#heaptype).

    -  slide 6 from https://www.blackhat.com/docs/us-16/materials/us-16-Yason-Windows-10-Segment-Heap-Internals.pdf (accompanying video: https://www.youtube.com/watch?v=hetZx78SQ_A). Additionally, this new heap implementation is the default for Modern apps.
    
If this is not enough evidence for you, I don't know what is.

> Also for the record, I won't approve it as the newly added attributes nullify the tuning knob here: https://github.com/qbittorrent/qBittorrent/blob/master/dist/windows/qt.conf#L5 and there exists other concerns I had in mind.

That tuning knob is commented-out by default for a reason - it's an old hack that for most does more harm than good. Previously, it was the best way we had to improve things in some corner cases, but this is no longer the case. Above I linked confirmation that **these changes to the manifest fixes real high DPI issues reported by users**, what more proof do you need?

Read through the docs, the new defaults are better than whatever we had before. Setting `dpiawareness=1` simply forces the application to be `System-DPI Aware` which is just worse than Per-Monitor (and Per-Monitor V2, available). The proven best way to deal with High DPI is to declare our application as PMv2/PM aware and let Qt handle the rest on its own (this has been discussed extensively at https://github.com/qbittorrent/qBittorrent/issues/13675 and other related issues).

> and there exists other concerns I had in mind.

State them, please. Otherwise, quips like these are nothing more than FUD.

Chocobo1 commented:
>But just because it isn't written that way word-for-word right now, doesn't mean my assertion is wrong. 

You are trying to **assert** something, you said it right there. It is either MS words or your words, I don't believe this is so hard to differentiate.

>That tuning knob is commented-out by default for a reason - it's an old hack that for most does more harm than good. 

I recall some users actually use it, but I care less since any further complaints is now on you.

>Otherwise, quips like these are nothing more than FUD.

I don't deny, but the same applies for your gospel of ["changing code page will make a difference for qbt"](https://github.com/qbittorrent/qBittorrent/pull/15020#discussion_r640122808).

>State them, please. 

How does dpi awareness relates to last screen position?
The mentioned issue doesn't state the resolution of the screens and they might be the same. One would be rationally suspecting that it is not the new additions that fixes it but other factors that are in play. One way to scientifically test is to provide 2 binaries from the same codebase (git master branch), one with the new additions and one without, although I suspect you will say that it is unnecessary to follow rigor scientifically tests.

FranciscoPombal commented:
@Chocobo1 

> You are trying to **assert** something, you said it right there. It is either MS words or your words, I don't believe this is so hard to differentiate.

I'm not asserting myself that one is recommended over the other. I was referring to _my assertion that **Microsoft (implicitly) recommends it**_; in other words, _I assert that Microsoft recommends it_. It's MS's words, not my words.

> Maybe MS did it on purpose.

What is this even supposed to mean? Are you suggesting that they introduced a new heap implementation, used it for critical systems components, advertised that they are now using it in Edge to great benefit, pointed to where developers can learn more about it and enable it themselves, only to fall short of an "actual recommendation" by not explicitly writing "this is recommended" in one sentence in the docs _on purpose_ as some sort of "signal" that it is still not production ready and should be avoided to enlightened individuals such as yourself? We need to hang out sometime, I love some juicy conspiracy theories.

> > That tuning knob is commented-out by default for a reason - it's an old hack that for most does more harm than good.
> 
> I recall some users actually use it, but I care less since any further complaints is now on you.

And any further praise as well :) (such as https://github.com/qbittorrent/qBittorrent/issues/15019#issuecomment-851013929)

> How does dpi awareness relates to last screen position?
> The mentioned issue doesn't state the resolution of the screens and they might be the same.

I don't know, I don't care, and neither should you, if you want to be pragmatic. High DPI is hard to get right. We should just be thankful that finally there are a couple of words that we can now put in the manifest and have it "just work" better than it ever did in the past, provided of course that Qt also does its part (which is the case now in recent versions, and is the point of a cross-platform GUI framework - to abstract all of this stuff away from us; also, it should be mentioned that this can only work 100% well in multi-monitor setups on recent Windows 10 versions where PMv2 is implemented, nothing can be done about problems arising from not having PMv2 on older versions).

If there are any outstanding issues that can't be resolved by simply applying these changes to the manifest or upgrading to a newer Qt version that would claim to fix such issues, then we can:

- Report the issue upstream and help upstream fix it.
- Get our hands dirty and write some sort of workaround ourselves (which may include fiddling with the options in qt.conf, as in the past).

But if not, then we don't have to worry about any of that.

> One would be rationally suspecting that it is not the new additions that fixes it but other factors that are in play. One way to scientifically test it is to provide 2 binaries from the same codebase (git master branch or v4_3_x), one with the new additions and one without, 

> although I suspect you will say that it is unnecessary to follow rigor scientifically tests and just hand wave it away..

Fuck your bullshit insults and FUD.
Go look at the discussions in https://github.com/qbittorrent/qBittorrent/projects/2#column-12115743, It's not just me, @xavier2k6 can tell you just how important these manifest attributes are for improving the High-DPI situation. You haven't been paying attention because you don't care. If you are so preoccupied you would have already taken the time to express such concerns earlier, engage in the discussion of the mentioned issues, and/or perhaps even setup such a rigorous experiment yourself. But that's not the goal now, is it? It's simply to harass me and stonewall my PRs with FUD.

 I'm actually interested in the improvement of this project. I have good reasons to doubt any commit between 4.3.5 and the latest `master` commit at the time https://github.com/qbittorrent/qBittorrent/issues/15019 was posted could have affected this.

Still, this is a good opportunity to dismiss the FUD and settle this once and for all. So here you go, I've done your work for you: https://github.com/qbittorrent/qBittorrent/issues/15019#issuecomment-870698237. I'll be happy with the result either way, even if I'm proven wrong - I agree with the premise of being as scientific as possible with such matters of course.

The takeaway here for you, is, that if you have such big concerns, take action yourself, engage in the relevant discussions, and be faster/more direct when stating your concerns instead of FUDding (posting Fermat-esque "I have truly grave concerns about this change which this post is too small to contain." posts) and insulting others claiming that they don't care about scientific rigor when they clearly do.

glassez commented:
I'm interested in one thing...
If all these "magic" fixes/improvements would certainly be good in any case, why wouldn't MS apply them without any participation from the application software developers (i.e. without any settings in manifest file)?
It looks more like they either have unproven effectiveness and are therefore provided as an option, or they do not fix anything by themselves, but only serve as a tool for implementing something.

sledgehammer999 commented:
@glassez probably legacy apps. AFAIK microsoft's whole history is going at great lengths ensuring old applications continue to work without a hitch in newer Windows versions. Hence, they can't auto enable new features for apps that don't declare they are compatible with them(the new features).

Chocobo1 commented:
>If all these "magic" fixes/improvements would certainly be good in any case, why wouldn't MS apply them without any participation from the application software developers (i.e. without any settings in manifest file)?
>
>@glassez probably legacy apps. 

Not just legacy apps might have trouble: https://www.windowslatest.com/2020/09/14/status-of-chrome-segment-heap-for-windows-10/
I'm didn't found any new follow-up, not sure how it ends.

FranciscoPombal commented:
> @glassez probably legacy apps. AFAIK microsoft's whole history is going at great lengths ensuring old applications continue to work without a hitch in newer Windows versions. Hence, they can't auto enable new features for apps that don't declare they are compatible with them(the new features).

This. We've been over this previously in other matters, such as with enabling `/Zc:__cplusplus`: also a straight up improvement and recommended best practice, but not enabled by default because of legacy apps (like really, really old stuff).

@Chocobo1 

> Not just legacy apps might have trouble, here the chrome case (which attribute is proposed in this PR):
> 
> * https://www.windowslatest.com/2020/09/14/status-of-chrome-segment-heap-for-windows-10/
> 
> * https://www.zdnet.com/article/google-disables-windows-10-ram-reduction-in-chrome/
> 
> * [Segment heap opt-in/opt-out flags needed microsoft/Windows-Dev-Performance#39](https://github.com/microsoft/Windows-Dev-Performance/issues/39)
> 
> 
> ~I didn't found any new follow-up, not sure how it ends.~ Seems they [didn't use the new segment heap and chose their own implementation](https://bugs.chromium.org/p/chromium/issues/detail?id=1102281#c30) in the end.

We are not a legacy app and we are not Chrome. Chrome is a very particular case due its size and complexity.
Web browsers today are the most complex and feature-crept pieces of software in the world, so I'd say it is not that unlikely that at least one of them would hit some very tricky corner case of this new improvement that actually causes a regression: [1](https://drewdevault.com/2020/03/18/Reckless-limitless-scope.html), [2](https://paste.sr.ht/~sircmpwn/13c1951014a256e9f551296a129bf6d10e9303dc).

As mentioned in https://github.com/qbittorrent/qBittorrent/pull/15020#issuecomment-870476414, even Microsoft has long since started using the new heap implementation for critical Windows components. Also, they are using it in Edge (which is based on Chromium), so presumably whatever was causing the regression in Chrome was fixed by Microsoft on their own fork... of course Chrome is a massive project with a ton of engineering resource, they can afford to roll their own solution.

xavier2k6 commented:
@Chocobo1 Another example of how the `dpiaware/dpiawareness` attributes help -> https://github.com/qbittorrent/qBittorrent/issues/15302?_pjax=%23repo-content-pjax-container#issuecomment-895138362

I created the build from a copy of the `v4_3_x Branch` & added the attributes to the `manifest` file.

xavier2k6 commented:
BTW, Just for the record - I have 3x 4K Monitor setup myself but I didn't/don't want to impose any bias.

Other users have now (twice anyway) shown that the dpi attributes help.

glassez commented:
@xavier2k6 
In any case, each problem should be considered separately, and not create such a dump of unrelated "improvements". Otherwise some one dubious can prevent others from being finished/merged.
